src/main/java/com/microsoft/sqlserver/jdbc/NTLMAuthentication.java

```diff
@@ -0,0 +1,846 @@
+        }
+
+        // verify message type
+        int messageType = token.getInt();
+        if (messageType != NTLM_MESSAGE_TYPE_CHALLENGE) {
+            MessageFormat form = new MessageFormat(SQLServerException.getErrString("R_ntlmMessageTypeError"));
+            Object[] msgArgs = {messageType};
+            throw new SQLServerException(form.format(msgArgs), null);
+        }
+
+        // target name fields
+        int targetNameLen = token.getShort();
+        token.getShort(); // targetNameMaxLen
+        token.getInt(); // targetNameOffset
+
+        // negotiate flags
+        token.getInt();
+
+        // get server challenge
+        token.get(context.serverChallenge);
+
+        // 8 bytes reserved 32-40: all 0's
+        token.getLong();
+
+        // target info fields
+        int targetInfoLen = token.getShort();
+        token.getShort(); // targetInfoMaxLen
+        token.getInt(); // targetInfoOffset
+
+        token.getLong(); // version - not used, for debug only
+
+        // get requested target name
+        byte[] targetName = new byte[targetNameLen];
+        token.get(targetName);
+
+        // verify targetInfo - was requested so should always be sent
+        context.targetInfo = new byte[targetInfoLen];
+        token.get(context.targetInfo);
+        if (0 == context.targetInfo.length) {
+            throw new SQLServerException(SQLServerException.getErrString("R_ntlmNoTargetInfo"), null);
+        }
+
+        // parse target info AV pairs
+        ByteBuffer targetInfoBuf = ByteBuffer.wrap(context.targetInfo).order(ByteOrder.LITTLE_ENDIAN);
+        boolean done = false;
+        for (int i = 0; i < context.targetInfo.length && !done;) {
+            int id = targetInfoBuf.getShort();
+            byte[] value = new byte[targetInfoBuf.getShort()];
+            targetInfoBuf.get(value);
+            switch (id) {
+                case NTLM_AVID_MSVAVTIMESTAMP:
+                    if (value.length > 0) {
+                        context.timestamp = new byte[NTLM_TIMESTAMP_LENGTH];
+                        System.arraycopy(value, 0, context.timestamp, 0, NTLM_TIMESTAMP_LENGTH);
+                    }
+                    break;
+                case NTLM_AVID_MSVAVEOL:
+                    done = true;
+                    break;
+                case NTLM_AVID_MSVAVDNSDOMAINNAME:
+                case NTLM_AVID_MSVAVDNSCOMPUTERNAME:
+                case NTLM_AVID_MSVAVFLAGS:
+                case NTLM_AVID_MSVAVNBCOMPUTERNAME:
+                case NTLM_AVID_MSVAVNBDOMAINNAME:
+                case NTLM_AVID_MSVAVDNSTREENAME:
+                case NTLM_AVID_MSVAVSINGLEHOST:
+                case NTLM_AVID_MSVAVTARGETNAME:
+                    break;
+                default:
+                    MessageFormat form = new MessageFormat(SQLServerException.getErrString("R_ntlmUnknownValue"));
+                    Object[] msgArgs = {new String(value, java.nio.charset.StandardCharsets.UTF_16LE)};

```

Comment:
Again, converting to String doesn't make much sense here. Just print out the token.